File: windoes/Nexus/nexus_flutter/lib/services/database_service.dart
import 'package:sqflite/sqflite.dart';
import 'package:sqflite_common_ffi/sqflite_ffi.dart';
import 'package:path/path.dart';
import 'dart:io';
import 'mongodb_service.dart';

class Note {
  int? id;
  String title;
  String content;
  DateTime createdAt;
  DateTime updatedAt;

  Note({
    this.id,
    required this.title,
    required this.content,
    required this.createdAt,
    required this.updatedAt,
  });

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'title': title,
      'content': content,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  factory Note.fromMap(Map<String, dynamic> map) {
    return Note(
      id: map['id'],
      title: map['title'],
      content: map['content'],
      createdAt: DateTime.parse(map['createdAt']),
      updatedAt: DateTime.parse(map['updatedAt']),
    );
  }
}

class Link {
  final int sourceId;
  final int targetId;

  Link({required this.sourceId, required this.targetId});
}

class GraphData {
  final List<Note> nodes;
  final List<Link> links;

  GraphData({required this.nodes, required this.links});
}

class DatabaseService {
  Database? _database;
  final MongoDbService _mongoService = MongoDbService();
  Function? onSyncComplete;

  Future<void> initialize() async {
    // Initialize FFI for desktop platforms
    if (Platform.isWindows || Platform.isLinux || Platform.isMacOS) {
      sqfliteFfiInit();
      databaseFactory = databaseFactoryFfi;
    }

    final dbPath = await getDatabasesPath();
    final path = join(dbPath, 'nexus.db');

    _database = await openDatabase(
      path,
      version: 1,
      onCreate: (db, version) async {
        await db.execute('''
          CREATE TABLE Notes (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            title TEXT NOT NULL,
            content TEXT NOT NULL,
            createdAt TEXT NOT NULL,
            updatedAt TEXT NOT NULL
          )
        ''');
        await db.execute('''
          CREATE TABLE Links (
            sourceId INTEGER,
            targetId INTEGER,
            PRIMARY KEY (sourceId, targetId)
          )
        ''');
      },
    );
    
    // Connect to MongoDB and sync
    await _mongoService.connect();
    
    // Set up auto-sync callback
    _mongoService.onDataChanged = () async {
      await syncFromMongoDB();
      if (onSyncComplete != null) {
        onSyncComplete!();
      }
    };
    
    await syncFromMongoDB();
  }

  Future<void> syncFromMongoDB() async {
    try {
      // Get all notes from MongoDB
      final mongoNotes = await _mongoService.getAllNotes();
      final mongoLinks = await _mongoService.getAllLinks();

      bool hasChanges = false;

      // Sync notes
      for (var mongoNote in mongoNotes) {
        final localId = mongoNote['localId'] as int;
        
        // Handle DateTime conversion
        String createdAt;
        String updatedAt;
        
        if (mongoNote['createdAt'] is DateTime) {
          createdAt = (mongoNote['createdAt'] as DateTime).toIso8601String();
        } else {
          createdAt = mongoNote['createdAt'] as String;
        }
        
        if (mongoNote['updatedAt'] is DateTime) {
          updatedAt = (mongoNote['updatedAt'] as DateTime).toIso8601String();
        } else {
          updatedAt = mongoNote['updatedAt'] as String;
        }
        
        final existing = await _database!.query(
          'Notes',
          where: 'id = ?',
          whereArgs: [localId],
        );

        if (existing.isEmpty) {
          // Insert new note
          await _database!.insert(
            'Notes',
            {
              'id': localId,
              'title': mongoNote['title'],
              'content': mongoNote['content'],
              'createdAt': createdAt,
              'updatedAt': updatedAt,
            },
            conflictAlgorithm: ConflictAlgorithm.replace,
          );
          print('✅ Inserted note: ${mongoNote['title']}');
          hasChanges = true;
        } else {
          // Update if MongoDB version is newer
          final localUpdated = DateTime.parse(existing.first['updatedAt'] as String);
          final mongoUpdated = DateTime.parse(updatedAt);
          
          if (mongoUpdated.isAfter(localUpdated)) {
            await _database!.update(
              'Notes',
              {
                'title': mongoNote['title'],
                'content': mongoNote['content'],
                'updatedAt': updatedAt,
              },
              where: 'id = ?',
              whereArgs: [localId],
            );
            print('✅ Updated note: ${mongoNote['title']}');
            hasChanges = true;
          }
        }
      }

      // Sync links
      for (var mongoLink in mongoLinks) {
        final result = await _database!.insert(
          'Links',
          {
            'sourceId': mongoLink['sourceId'],
            'targetId': mongoLink['targetId'],
          },
          conflictAlgorithm: ConflictAlgorithm.ignore,
        );
        if (result > 0) hasChanges = true;
      }
      
      if (hasChanges) {
        print('🔄 Changes detected - triggering UI update');
      }
    } catch (e) {
      print('MongoDB sync error: $e');
      rethrow;
    }
  }

  Future<List<Note>> getAllNotes() async {
    final List<Map<String, dynamic>> maps = await _database!.query(
      'Notes',
      orderBy: 'updatedAt DESC',
    );
    return List.generate(maps.length, (i) => Note.fromMap(maps[i]));
  }

  Future<void> saveNote(Note note) async {
    if (note.id == null) {
      note.id = await _database!.insert('Notes', note.toMap());
    } else {
      await _database!.update(
        'Notes',
        note.toMap(),
        where: 'id = ?',
        whereArgs: [note.id],
      );
    }
    await _updateLinks(note);
    
    // Sync to MongoDB
    await _mongoService.syncNote(note);
  }

  Future<void> _updateLinks(Note note) async {
    await _database!.delete('Links', where: 'sourceId = ?', whereArgs: [note.id]);
    
    final linkPattern = RegExp(r'\[\[([^\]]+)\]\]');
    final matches = linkPattern.allMatches(note.content);
    
    for (final match in matches) {
      final targetTitle = match.group(1);
      final targetNotes = await _database!.query(
        'Notes',
        where: 'title = ?',
        whereArgs: [targetTitle],
      );
      
      if (targetNotes.isNotEmpty) {
        await _database!.insert(
          'Links',
          {'sourceId': note.id, 'targetId': targetNotes.first['id']},
          conflictAlgorithm: ConflictAlgorithm.ignore,
        );
      }
    }
  }

  Future<void> deleteNote(int id) async {
    await _database!.delete('Links', where: 'sourceId = ? OR targetId = ?', whereArgs: [id, id]);
    await _database!.delete('Notes', where: 'id = ?', whereArgs: [id]);
    
    // Sync to MongoDB
    await _mongoService.deleteNote(id);
  }

  Future<List<Note>> search(String query) async {
    final List<Map<String, dynamic>> maps = await _database!.query(
      'Notes',
      where: 'title LIKE ? OR content LIKE ?',
      whereArgs: ['%$query%', '%$query%'],
      orderBy: 'updatedAt DESC',
    );
    return List.generate(maps.length, (i) => Note.fromMap(maps[i]));
  }

  Future<GraphData> getGraphData() async {
    final notes = await getAllNotes();
    final List<Map<String, dynamic>> linkMaps = await _database!.query('Links');
    final links = linkMaps.map((m) => Link(sourceId: m['sourceId'], targetId: m['targetId'])).toList();
    return GraphData(nodes: notes, links: links);
  }

  Future<void> createLink(int sourceId, int targetId) async {
    await _database!.insert(
      'Links',
      {'sourceId': sourceId, 'targetId': targetId},
      conflictAlgorithm: ConflictAlgorithm.ignore,
    );
    
    // Sync to MongoDB
    await _mongoService.syncLink(sourceId, targetId);
  }

  Future<void> deleteLink(int sourceId, int targetId) async {
    await _database!.delete(
      'Links',
      where: 'sourceId = ? AND targetId = ?',
      whereArgs: [sourceId, targetId],
    );
    
    // Sync to MongoDB
    await _mongoService.deleteLink(sourceId, targetId);
  }

  Future<List<Link>> getAllLinks() async {
    final List<Map<String, dynamic>> maps = await _database!.query('Links');
    return maps.map((m) => Link(sourceId: m['sourceId'], targetId: m['targetId'])).toList();
  }
}

File: windoes/Nexus/Nexus/Services/DatabaseService.cs
using Microsoft.Data.Sqlite;
using Nexus.Models;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.RegularExpressions;

namespace Nexus.Services;

public class DatabaseService
{
    private readonly string dbPath;
    private readonly string connectionString;
    private readonly MongoDbService _mongoService;

    public MongoDbService MongoService => _mongoService;

    public DatabaseService()
    {
        var appData = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
        var folder = Path.Combine(appData, "Nexus");
        Directory.CreateDirectory(folder);
        dbPath = Path.Combine(folder, "nexus.db");
        connectionString = $"Data Source={dbPath}";
        _mongoService = new MongoDbService();
        InitializeDatabase();
    }

    private void InitializeDatabase()
    {
        using var conn = new SqliteConnection(connectionString);
        conn.Open();

        var cmd = conn.CreateCommand();
        cmd.CommandText = @"
            CREATE TABLE IF NOT EXISTS Notes (
                Id INTEGER PRIMARY KEY AUTOINCREMENT,
                Title TEXT NOT NULL,
                Content TEXT NOT NULL,
                CreatedAt TEXT NOT NULL,
                UpdatedAt TEXT NOT NULL
            );
            CREATE VIRTUAL TABLE IF NOT EXISTS NotesSearch USING fts5(Title, Content);
            CREATE TABLE IF NOT EXISTS Links (
                SourceId INTEGER,
                TargetId INTEGER,
                PRIMARY KEY (SourceId, TargetId)
            );";
        cmd.ExecuteNonQuery();
    }

    public void DeleteLink(int sourceId, int targetId)
    {
        using var conn = new SqliteConnection(connectionString);
        conn.Open();
        
        var cmd = conn.CreateCommand();
        cmd.CommandText = "DELETE FROM Links WHERE SourceId=@source AND TargetId=@target";
        cmd.Parameters.AddWithValue("@source", sourceId);
        cmd.Parameters.AddWithValue("@target", targetId);
        cmd.ExecuteNonQuery();
        
        // Sync to MongoDB
        _ = _mongoService.DeleteLinkAsync(sourceId, targetId);
    }

    public List<Link> GetAllLinks()
    {
        var links = new List<Link>();
        using var conn = new SqliteConnection(connectionString);
        conn.Open();
        
        var cmd = conn.CreateCommand();
        cmd.CommandText = "SELECT SourceId, TargetId FROM Links";
        using var reader = cmd.ExecuteReader();
        while (reader.Read())
            links.Add(new Link { SourceId = reader.GetInt32(0), TargetId = reader.GetInt32(1) });
        
        return links;
    }

    public void CreateLink(int sourceId, int targetId)
    {
        using var conn = new SqliteConnection(connectionString);
        conn.Open();
        
        var cmd = conn.CreateCommand();
        cmd.CommandText = "INSERT OR IGNORE INTO Links (SourceId, TargetId) VALUES (@source, @target)";
        cmd.Parameters.AddWithValue("@source", sourceId);
        cmd.Parameters.AddWithValue("@target", targetId);
        cmd.ExecuteNonQuery();
        
        // Sync to MongoDB
        _ = _mongoService.SyncLinkAsync(sourceId, targetId);
    }

    public void DeleteNote(int noteId)
    {
        using var conn = new SqliteConnection(connectionString);
        conn.Open();
        
        using var transaction = conn.BeginTransaction();
        try
        {
            // Delete links
            var delLinksCmd = conn.CreateCommand();
            delLinksCmd.CommandText = "DELETE FROM Links WHERE SourceId=@id OR TargetId=@id";
            delLinksCmd.Parameters.AddWithValue("@id", noteId);
            delLinksCmd.ExecuteNonQuery();
            
            // Delete note
            var delNoteCmd = conn.CreateCommand();
            delNoteCmd.CommandText = "DELETE FROM Notes WHERE Id=@id";
            delNoteCmd.Parameters.AddWithValue("@id", noteId);
            delNoteCmd.ExecuteNonQuery();
            
            transaction.Commit();
            
            // Sync to MongoDB
            _ = _mongoService.DeleteNoteAsync(noteId);
        }
        catch
        {
            transaction.Rollback();
            throw;
        }
    }

    public List<Note> Search(string query)
    {
        var results = new List<Note>();
        using var conn = new SqliteConnection(connectionString);
        conn.Open();
        
        var cmd = conn.CreateCommand();
        cmd.CommandText = @"
            SELECT n.* FROM Notes n
            WHERE n.Title LIKE @query OR n.Content LIKE @query
            ORDER BY n.UpdatedAt DESC";
        cmd.Parameters.AddWithValue("@query", $"%{query}%");
        
        using var reader = cmd.ExecuteReader();
        while (reader.Read())
        {
            results.Add(new Note
            {
                Id = reader.GetInt32(0),
                Title = reader.GetString(1),
                Content = reader.GetString(2),
                CreatedAt = DateTime.Parse(reader.GetString(3)),
                UpdatedAt = DateTime.Parse(reader.GetString(4))
            });
        }
        return results;
    }

    public List<Note> GetAllNotes()
    {
        var notes = new List<Note>();
        using var conn = new SqliteConnection(connectionString);
        conn.Open();
        var cmd = conn.CreateCommand();
        cmd.CommandText = "SELECT * FROM Notes ORDER BY UpdatedAt DESC";
        using var reader = cmd.ExecuteReader();
        while (reader.Read())
        {
            notes.Add(new Note
            {
                Id = reader.GetInt32(0),
                Title = reader.GetString(1),
                Content = reader.GetString(2),
                CreatedAt = DateTime.Parse(reader.GetString(3)),
                UpdatedAt = DateTime.Parse(reader.GetString(4))
            });
        }
        return notes;
    }

    public void SaveNote(Note note)
    {
        using var conn = new SqliteConnection(connectionString);
        conn.Open();
        
        using var transaction = conn.BeginTransaction();
        try
        {
            var cmd = conn.CreateCommand();
            
            if (note.Id == 0)
            {
                cmd.CommandText = "INSERT INTO Notes (Title, Content, CreatedAt, UpdatedAt) VALUES (@title, @content, @created, @updated); SELECT last_insert_rowid();";
                cmd.Parameters.AddWithValue("@title", note.Title);
                cmd.Parameters.AddWithValue("@content", note.Content);
                cmd.Parameters.AddWithValue("@created", note.CreatedAt.ToString("o"));
                cmd.Parameters.AddWithValue("@updated", note.UpdatedAt.ToString("o"));
                note.Id = Convert.ToInt32(cmd.ExecuteScalar());
            }
            else
            {
                cmd.CommandText = "UPDATE Notes SET Title=@title, Content=@content, UpdatedAt=@updated WHERE Id=@id";
                cmd.Parameters.AddWithValue("@id", note.Id);
                cmd.Parameters.AddWithValue("@title", note.Title);
                cmd.Parameters.AddWithValue("@content", note.Content);
                cmd.Parameters.AddWithValue("@updated", note.UpdatedAt.ToString("o"));
                cmd.ExecuteNonQuery();
            }
            
            UpdateLinks(note, conn);
            transaction.Commit();
            
            // Sync to MongoDB
            _ = _mongoService.SyncNoteAsync(note);
        }
        catch
        {
            transaction.Rollback();
            throw;
        }
    }

    private void UpdateLinks(Note note, SqliteConnection conn)
    {
        var links = ExtractLinks(note.Content);
        
        var delCmd = conn.CreateCommand();
        delCmd.CommandText = "DELETE FROM Links WHERE SourceId=@id";
        delCmd.Parameters.AddWithValue("@id", note.Id);
        delCmd.ExecuteNonQuery();

        foreach (var link in links)
        {
            var targetCmd = conn.CreateCommand();
            targetCmd.CommandText = "SELECT Id FROM Notes WHERE Title=@title";
            targetCmd.Parameters.AddWithValue("@title", link);
            var targetId = targetCmd.ExecuteScalar();
            
            if (targetId != null)
            {
                var insCmd = conn.CreateCommand();
                insCmd.CommandText = "INSERT OR IGNORE INTO Links (SourceId, TargetId) VALUES (@source, @target)";
                insCmd.Parameters.AddWithValue("@source", note.Id);
                insCmd.Parameters.AddWithValue("@target", targetId);
                insCmd.ExecuteNonQuery();
            }
        }
    }

    private List<string> ExtractLinks(string content)
    {
        var links = new List<string>();
        var matches = Regex.Matches(content, @"\[\[([^\]]+)\]\]");
        foreach (Match match in matches)
            links.Add(match.Groups[1].Value);
        return links;
    }

    public GraphData GetGraphData()
    {
        var notes = GetAllNotes();
        var links = new List<Link>();
        
        using var conn = new SqliteConnection(connectionString);
        conn.Open();
        var cmd = conn.CreateCommand();
        cmd.CommandText = "SELECT SourceId, TargetId FROM Links";
        using var reader = cmd.ExecuteReader();
        while (reader.Read())
            links.Add(new Link { SourceId = reader.GetInt32(0), TargetId = reader.GetInt32(1) });

        return new GraphData { Nodes = notes.ToArray(), Links = links.ToArray() };
    }

    public async System.Threading.Tasks.Task SyncFromMongoDBAsync()
    {
        try
        {
            // Get all notes from MongoDB
            var mongoNotes = await _mongoService.GetAllNotesAsync();
            var mongoLinks = await _mongoService.GetAllLinksAsync();

            using var conn = new SqliteConnection(connectionString);
            conn.Open();

            // Sync notes
            foreach (var mongoNote in mongoNotes)
            {
                var cmd = conn.CreateCommand();
                cmd.CommandText = "SELECT Id FROM Notes WHERE Id=@id";
                cmd.Parameters.AddWithValue("@id", mongoNote.LocalId);
                var exists = cmd.ExecuteScalar();

                if (exists == null)
                {
                    // Insert new note
                    var insertCmd = conn.CreateCommand();
                    insertCmd.CommandText = @"INSERT INTO Notes (Id, Title, Content, CreatedAt, UpdatedAt) 
                                             VALUES (@id, @title, @content, @created, @updated)";
                    insertCmd.Parameters.AddWithValue("@id", mongoNote.LocalId);
                    insertCmd.Parameters.AddWithValue("@title", mongoNote.Title);
                    insertCmd.Parameters.AddWithValue("@content", mongoNote.Content);
                    insertCmd.Parameters.AddWithValue("@created", mongoNote.CreatedAt.ToString("o"));
                    insertCmd.Parameters.AddWithValue("@updated", mongoNote.UpdatedAt.ToString("o"));
                    insertCmd.ExecuteNonQuery();
                }
                else
                {
                    // Update existing note if MongoDB version is newer
                    var updateCmd = conn.CreateCommand();
                    updateCmd.CommandText = @"UPDATE Notes 
                                             SET Title=@title, Content=@content, UpdatedAt=@updated 
                                             WHERE Id=@id AND UpdatedAt < @updated";
                    updateCmd.Parameters.AddWithValue("@id", mongoNote.LocalId);
                    updateCmd.Parameters.AddWithValue("@title", mongoNote.Title);
                    updateCmd.Parameters.AddWithValue("@content", mongoNote.Content);
                    updateCmd.Parameters.AddWithValue("@updated", mongoNote.UpdatedAt.ToString("o"));
                    updateCmd.ExecuteNonQuery();
                }
            }

            // Sync links
            foreach (var mongoLink in mongoLinks)
            {
                var cmd = conn.CreateCommand();
                cmd.CommandText = "INSERT OR IGNORE INTO Links (SourceId, TargetId) VALUES (@source, @target)";
                cmd.Parameters.AddWithValue("@source", mongoLink.SourceId);
                cmd.Parameters.AddWithValue("@target", mongoLink.TargetId);
                cmd.ExecuteNonQuery();
            }
        }
        catch (Exception ex)
        {
            System.Diagnostics.Debug.WriteLine($"MongoDB sync error: {ex.Message}");
        }
    }
}
